### PR TITLE
[infra] Reopen - Create API for individual team use

### PR DESCRIPTION
**Pull request by Razvan00Rusu**
Restored #39 branch and reopened PR to merge later.

### CODE REVIEW

**Review by sherwin-dc — APPROVED** [review]

**Comment by neeldug** [comment]
Just quick question regarding this PR, what exactly does this add with respect to functionality as opposed to an agent team spawning in agents with their own given behaviours?

**Comment by neeldug** [comment]
> Just quick question regarding this PR, what exactly does this add with respect to functionality as opposed to an agent team spawning in agents with their own given behaviours?

Also, I do think the refactor is great, I just don't see the necessity of adding makefile options on top of environment variables

**Comment by sherwin-dc** [comment]
> Just quick question regarding this PR, what exactly does this add with respect to functionality as opposed to an agent team spawning in agents with their own given behaviours?

Will let me add agent maps in team files in the next PR

File: pkg/infra/game/stage/fight/fight.go
Comment (neeldug) [inline pkg/infra/game/stage/fight/fight.go:36]:
```suggestion
func AgentFightDecisions(state *state.View, agents map[commons.ID]agent.Agent, previousDecisions *immutable.Map[commons.ID, decision.FightAction], channelsMap map[commons.ID]chan message.TaggedMessage) map[commons.ID]decision.FightAction {
```

**Comment by sherwin-dc** [comment]
> > Just quick question regarding this PR, what exactly does this add with respect to functionality as opposed to an agent team spawning in agents with their own given behaviours?
> 
> Also, I do think the refactor is great, I just don't see the necessity of adding makefile options on top of environment variables

I dont think there is another way of say, calling a different initialise() function that another team has written. I think functions can be called by names, but then all those functions need to be in the same package (unless I am mistaken on this)

File: pkg/infra/game/stage/fight/fight.go
Comment (neeldug) [inline pkg/infra/game/stage/fight/fight.go:113]: Why is this now a pointer to decisionLog? I think this will become racy since unsafe stuff can happen between goroutines, i.e. writing directly to the same pointer.

**Review by neeldug — CHANGES_REQUESTED** [review]

**Comment by neeldug** [comment]
> I dont think there is another way of say, calling a different initialise() function that another team has written. I think functions can be called by string names, but then all those functions need to be in the same package (unless I am mistaken on this)

Sure, I think this is fine, but just to make sure - any agent team can still run without specifying anything in the previous fashion to still run the game as is?

**Comment by sherwin-dc** [comment]
> > > Just quick question regarding this PR, what exactly does this add with respect to functionality as opposed to an agent team spawning in agents with their own given behaviours?
> > 
> > 
> > Also, I do think the refactor is great, I just don't see the necessity of adding makefile options on top of environment variables
> 
> I dont think there is another way of say, calling a different initialise() function that another team has written. I think functions can be called by string names, but then all those functions need to be in the same package (unless I am mistaken on this)

Oh wait if you had meant changing the string variable 'mode' using an environment variable then maybe it is possible instead

**Comment by sherwin-dc** [comment]
> > I dont think there is another way of say, calling a different initialise() function that another team has written. I think functions can be called by string names, but then all those functions need to be in the same package (unless I am mistaken on this)
> 
> Sure, I think this is fine, but just to make sure - any agent team can still run without specifying anything in the previous fashion to still run the game as is?

Yes, just running `make` will build the default game as it is now

File: pkg/infra/game/stage/fight/fight.go
Comment (neeldug) [inline pkg/infra/game/stage/fight/fight.go:72]:
```suggestion
func HandleFightRound(state *state.State, baseHealth uint, decisionMap map[commons.ID]decision.FightAction) (uint, uint, uint) {
```

File: pkg/infra/teams/team0/fight_decisions.go
Comment (neeldug) [inline pkg/infra/teams/team0/fight_decisions.go:9]:
```suggestion
func AllDefend(agents map[commons.ID]agent.Agent) map[commons.ID]decision.FightAction {
	decisionMap := make(map[commons.ID]decision.FightAction)
```

**Review by neeldug — CHANGES_REQUESTED** [review]

**Review by neeldug — APPROVED** [review]
LGTM, probably wait for a second reviewer before merging in case I missed something.

**Comment by sherwin-dc** [comment]
Yup done with adding changes for now. I'll move the initialise function to team files in a separate PR.

**Comment by Razvan00Rusu** [comment]
> LGTM, probably wait for a second reviewer before merging in case I missed something. Actually @Razvan00Rusu since you opened the PR I guess you can't formally review, is it good to merge?

Just had a look through it now,  the premise looks good and it all still works which is the main thing 👀. I'll merge this through and just bare it in mind if anything needs to be addressed in a new PR later.

@sherwin-dc, this is a big structural change though so it might be worth documenting somewhere even for the Infra team (GitHub wiki is fine) :D

**Pull request merged**
